gutenberg.php

```diff
@@ -44,30 +196,67 @@ function gutenberg_build_files_notice() {
 }
 
 /**
- * Verify that we can initialize the Gutenberg editor , then load it.
+ * Verify that we can initialize the Gutenberg editor, then load it.
  *
  * @since 1.5.0
  */
 function gutenberg_pre_init() {
-	global $wp_version;
-	if ( defined( 'GUTENBERG_DEVELOPMENT_MODE' ) && GUTENBERG_DEVELOPMENT_MODE && ! file_exists( __DIR__ . '/build/blocks' ) ) {
-		add_action( 'admin_notices', 'gutenberg_build_files_notice' );
-		return;
-	}
-
 	// Get unmodified $wp_version.
 	include ABSPATH . WPINC . '/version.php';
 
-	// Strip '-src' from the version string. Messes up version_compare().
-	$version = str_replace( '-src', '', $wp_version );
+	// Strip '-src' from the WP version string. May interfere with `version_compare()`.
+	$wp_version = str_replace( '-src', '', $wp_version );
+
+	/*
+	 * Note that `version_compare()` considers 1 < 1.0 < 1.0.0 (before suffixes)
+	 * so 6.0 < 6.0.0-alpha and 6.0-beta < 6.0.0-alpha are both true.
+	 * To work around any possible problems when comparing with WP development version strings
+	 * the GUTENBERG_MAX_WP_VERSION is incremented by one and the comparisons are with `>=`.
+	 */
+	$max_supported_version = (float) GUTENBERG_MAX_WP_VERSION + 0.1;
 
-	// Compare against major release versions (X.Y) rather than minor (X.Y.Z)
-	// unless a minor release is the actual minimum requirement. WordPress reports
-	// X.Y for its major releases.
-	if ( version_compare( $version, '5.9', '<' ) ) {
-		add_action( 'admin_notices', 'gutenberg_wordpress_version_notice' );
+	// `version_compare()` expects strings.
+	$max_supported_version = (string) $max_supported_version;
+
+	// Check if this version of Gutenberg supports the version of WordPress.
+	if ( version_compare( $wp_version, GUTENBERG_MIN_WP_VERSION, '<' ) ) {
+		add_action( 'admin_notices', 'gutenberg_wordpress_version_too_old_notice' );
+
+		// Also add a notice to the plugin's row in the plugins list tables.
+		add_filter( 'plugin_row_meta', 'gutenberg_wordpress_version_too_old_plugin_row_meta', 10, 2 );
+
+		// Return early, do not load Gutenberg.
+		return;
+	} elseif ( version_compare( $wp_version, $max_supported_version, '>=' ) ) {
+		/*
+		 * Do not load Gutenebrg in newer, unsupported versions of WordPress. This will prevent incompatibilities,
+		 * PHP fatal errors, etc. Ask the users to update Gutenberg or turn plugin auto-updates on.
+		*/
+		add_action( 'admin_notices', 'gutenberg_version_too_old_notice' );
+
+		// Also add a notice to the plugin's row in the plugins list table.
+		add_filter( 'plugin_row_meta', 'gutenberg_version_too_old_plugin_row_meta', 10, 2 );
+
+		return;
+	} elseif (
+		version_compare( $wp_version, $max_supported_version, '<' ) &&
+		version_compare( $wp_version, $max_supported_version . '-alpha', '>=' )
+	) {
```

Comment:
To test this I set the WP max version to 6.3 and I have 6.4 (beta or RC) installed locally, and I expected the plugin to disable entirely, but for now I just got that it will be disabled when I upgrade. To be honest, I think that's a bit confusing and I'd just disable Gutenberg in this case as well. I'm not sure we need the two levels of feedback.

Reply:
I went ahead and made this change. So now:

 - The plugin is also disabled during the beta/RC period. The reasoning is that the Gutenberg features and new code have already been back ported to Core at that time. And you need to update Gutenberg to the next version in order to include the features that were not back ported (targeted for the next WP version).
 - It also means that the max version should be updated right before/after beta 1.

Reply:
> It also means that the max version should be updated right before/after beta 1

Yep, as soon as the version with the latest fixes and features for the current WP development version is released (the latest Gutenberg version scheduled to be synced with core trunk). 

As far as I see further fixes to merged features are done in minor (dot) versions. There `GUTENBERG_MAX_WP_VERSION` would be same as in the (older) major version.